PyRDF/backend/Dist.py

```diff
@@ -226,55 +255,93 @@ def _get_clustered_ranges(self, nentries, treename, filelist,
             warnings.warn(msg, UserWarning, stacklevel=2)
             self.npartitions = numclusters
 
-        partSize = numclusters // self.npartitions
-        remainder = numclusters % self.npartitions
+        logger.debug("{} clusters will be split along {} partitions.".format(
+            numclusters, self.npartitions))
 
-        i = 0  # Iterator
-        ranges = []
-        entries_to_process = 0
-
-        while i < numclusters:
-            index_start = i
-            start = clusters[i][0]
-            i = i + partSize
-            if remainder > 0:
-                i += 1
-                remainder -= 1
-            index_end = i
-            if i == numclusters:
-                end = clusters[-1][1]
-            else:
-                end = clusters[i - 1][1]
-
-            range_files = []
-            for idx in range(index_start, index_end):
-                current_file = clusters[idx][3]
-                if range_files and range_files[-1] == current_file:
-                    continue
-                range_files.append(clusters[idx][3])
+        """
+        This list comprehension builds the elements of the ranged tuples:
+
+        1. ``start``: The minimum entry among all the clusters considered in a
+           given partition. The offset of the current file is subtracted
+        2. ``end``: The maximum entry among all the clusters considered in a
+           given partition. The offset of the current file is subtracted
+        3. ``filelist``: The list of files that are span between entries
+           ``start`` and ``end``::
+
+                Filelist: [file_1,file_2,file_3,file_4]
+
+                Clustered range: [0,150]
+                file_1 holds entries [0, 100]
+                file_2 holds entries [101, 200]
+                Then the clustered range should open [file_1, file_2]
+
+                Clustered range: [150,350]
+                file_3 holds entris [201, 300]
+                file_4 holds entries [301, 400]
+                Then the clustered range should open [file_2, file_3, file_4]
+
+           Each ``cluster`` namedtuple has a ``fileandindex`` namedtuple. The
+           second element of this tuple corresponds to the index of the file in
+           the input `TChain`. This way all files can be uniquely identified,
+           even if there is some repetition (e.g. when building a TChain with
+           multiple instances of the same file). The algorithm to retrieve the
+           correct files for each range takes the unique filenames from the list
+           of clusters and sorts them by their index to keep the original order.
+
+           In each file only the clusters needed to process the clustered range
+           will be read.
+        4. ``friend_info``: Information about friend trees.
+        """
+        clustered_ranges = [
+            Range(
+                min(clusters)[0] - clusters[0].offset,  # type: int
```

Comment:
Why always subtracting `clusters[0].offset` ?

Reply:
We need to subtract the offset of a given entry in the TChain. The offset is per-file based, meaning that the offset changes when the current file of the chain changes, i.e. the first offset is always zero and when the entries of the first file have been read it will be increased by the number of entries of that file and so on.

With that said, while min(clusters)[0] will always lead to the first cluster in `clusters`, thus clusters[0].offset is the correct offset for that particular entry, it may not be the same for max(clusters)[1], that will instead lead to the last cluster in `clusters`

Reply:
What I'm trying to say is that probably the next line should be
```suggestion
               max(clusters)[1] - clusters[-1].offset,  # type: int
```

Reply:
Yes, well, we need to make sure that both `min(clusters)[0]` and `max(clusters)[1]` get subtracted the offset of the file they belong to.
So `min(clusters)` is the lowest cluster in the range, and `max(clusters)` is the highest? It would be useful to add a comment next to these two lines, saying something like `start of the range: start of lowest cluster minus offset of its file` and `end of the range: end of the highest cluster minus offset of its file`.

Reply:
Actually revisiting the master code at https://github.com/vepadulano/PyRDF/blob/554b04f14341c815d7cdb9bd194af15cb1133620/PyRDF/backend/Dist.py#L270 the offset is always taken from the first cluster of the range, also for the `end` entry. The original PR didn't comment on the reason this is done, I suppose this has to do with the internals of TChain, I'm investigating

Reply:
I made a simple example to understand this better. Given the following:
```
tree10000entries10clusters.root --> 1 TFile with 1 TTree, 10000 entries, 10 clusters
tree20000entries10clusters.root --> 1 TFile with 1 TTree, 20000 entries, 10 clusters
tree30000entries10clusters.root --> 1 TFile with 1 TTree, 30000 entries, 10 clusters
```

**TOTAL ENTRIES TO READ FROM THE CHAIN = 60000** 

### Original logic to build partitions/ranges (using always the offset of the first cluster)
If we run the code to build 2 partitions/ranges with the master PyRDF logic we get

```
Range(start=0, end=20000, filelist=['tree10000entries10clusters.root', 'tree20000entries10clusters.root'], friend_info=None)

Range(start=10000, end=50000, filelist=['tree20000entries10clusters.root', 'tree30000entries10clusters.root'], friend_info=None)
```
So the first range will read 20000 entries [0,20000] , of which the first 10000 are read from `tree10000entries10clusters.root` and the other 10000 are read from `tree20000entries10clusters.root`
The second range instead will read 40000 entries [10000, 50000] (and these are local to the trees of the filelist of the second range), so the reading will start from entry 10000 of the tree in file `tree20000entries10clusters.root`, it will finish all entries in that file [10000,20000] then it will still have to read another 30000 entries (50000-20000) that are exactly the entries in file `tree30000entries10clusters.root`

**TOTAL ENTRIES READ = 60000**

### The other case proposed in the comments above
What happens if we take the offset of the last cluster (or any other cluster that belongs to a file different from the one of the first cluster) ?
```
Range(start=0, end=10000, filelist=['tree10000entries10clusters.root', 'tree20000entries10clusters.root'], friend_info=None)

Range(start=10000, end=30000, filelist=['tree20000entries10clusters.root', 'tree30000entries10clusters.root'], friend_info=None)
```
The first range will read only 10000 entries this time, since the offset of the second tree (which is indeed 10000) was subtracted to the `end` entry (20000-10000). So even though we have two files in the list of files for this range, we are only reading the entries of the first one.
For the second range even worse, a total of only 20000 entries is read [10000,30000]. The reading will start from entry 10000 of file `tree20000entries10clusters.root` up to the end of the file (entry 20000). Finally only another 10000 entries will be read from file `tree30000entries10clusters.root`

**TOTAL ENTRIES READ= 30000**

### Bottomline
only `clusters[0].offset` should be subtracted from either `start` or `end` entry to avoid losing data

Reply:
An extra: creating 4 ranges instead of 2 leads to even weirder results for the second case
```
Range(start=0, end=8000, filelist=['tree10000entries10clusters.root'], friend_info=None)

Range(start=8000, end=12000, filelist=['tree10000entries10clusters.root', 'tree20000entries10clusters.root'], friend_info=None)

Range(start=12000, end=9000, filelist=['tree20000entries10clusters.root', 'tree30000entries10clusters.root'], friend_info=None)

Range(start=9000, end=30000, filelist=['tree30000entries10clusters.root'], friend_info=None)
```
Notice how range 3 starts at 12000 and ends at 9000